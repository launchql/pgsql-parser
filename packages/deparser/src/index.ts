import { Deparser } from "./deparser";

const deparse = Deparser.deparse;
export { deparse, Deparser };
