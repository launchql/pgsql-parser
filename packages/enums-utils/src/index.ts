import toInt from './enums2int.json';
import toStr from './enums2str.json';
import nodes from './nodes.json';

export { toInt, toStr, nodes };
export const getEnum = (enumType, key) => {
  if (typeof key === 'number') {
    return toStr[enumType][key];
  } else {
    return toInt[enumType][key];
  }
};

export * from './types';
