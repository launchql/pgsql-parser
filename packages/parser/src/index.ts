import { Deparser, deparse } from 'pgsql-deparser';
import {
  parseQuery,
  parseQuerySync,
  parsePlPgSQLSync as parseFunction
} from 'libpg-query';

function mapStmt({ stmt, stmt_len, stmt_location }) {
  return {
    RawStmt: {
      stmt,
      stmt_len,
      stmt_location: stmt_location || 0
    }
  };
}

export const parse = (sql) => {
  if (!sql) throw new Error('no SQL provided to parser');
  const result = parseQuerySync(sql);
  return result.stmts.map(mapStmt);
};

export const parseAsync = async (sql) => {
  if (!sql) throw new Error('no SQL provided to parser');
  const result = await parseQuery(sql);
  return result.stmts.map(mapStmt);
};

export { deparse, Deparser, parseFunction };
