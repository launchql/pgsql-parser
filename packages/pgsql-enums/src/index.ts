import toInt from './enums2int';
import toStr from './enums2str';
import nodes from './nodes';

export { toInt, toStr, nodes };
export const getEnum = (enumType, key) => {
  if (typeof key === 'number') {
    return toStr[enumType][key];
  } else {
    return toInt[enumType][key];
  }
};

export * from './types';
