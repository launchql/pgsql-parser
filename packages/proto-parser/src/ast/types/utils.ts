import * as t from '@babel/types';
import { isPrimitiveType } from '../../utils';

export const getTSType = (type: string = 'any') => {
  switch (type) {
    case 'string':
      return t.tsStringKeyword();
    case 'double':
    case 'float':
    case 'int32':
    case 'uint32':
    case 'sint32':
    case 'fixed32':
    case 'sfixed32':
      return t.tsNumberKeyword();
    case 'int64':
    case 'uint64':
    case 'sint64':
    case 'fixed64':
    case 'sfixed64':
      return t.tsBigIntKeyword()
    case 'bytes':
      return t.tsTypeReference(t.identifier('Uint8Array'));
    case 'bool':
      return t.tsBooleanKeyword();
    default:
      // For custom types, reference them directly
      return t.tsTypeReference(t.identifier(type));

  };
};

export const resolveTypeName = (type: string) => {
  if (isPrimitiveType(type)) {
    return getTSType(type);
  }
  return t.tsTypeReference(t.identifier(type));
}
