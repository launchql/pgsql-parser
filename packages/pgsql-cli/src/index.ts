#!/usr/bin/env node
import minimist from 'minimist';
import chalk from 'chalk';
import { readFileSync } from 'fs';
import { resolve } from 'path';

import { parseCommand } from './commands/parse';
import { deparseCommand } from './commands/deparse';
import { protoGenCommand } from './commands/proto-gen';
import { protoFetchCommand } from './commands/proto-fetch';
import { runtimeSchemaCommand } from './commands/runtime-schema';
import { showHelp, showVersion } from './utils/help';

const argv = minimist(process.argv.slice(2), {
  alias: {
    h: 'help',
    v: 'version',
    o: 'output',
    f: 'format'
  }
});

const command = argv._[0];

async function main() {
  try {
    if (argv.version) {
      showVersion();
      process.exit(0);
    }

    if (!command || argv.help) {
      showHelp(command);
      process.exit(0);
    }

    switch (command) {
      case 'parse':
        await parseCommand(argv);
        break;
      
      case 'deparse':
        await deparseCommand(argv);
        break;
      
      case 'proto-gen':
        await protoGenCommand(argv);
        break;
      
      case 'proto-fetch':
        await protoFetchCommand(argv);
        break;
      
      case 'runtime-schema':
        await runtimeSchemaCommand(argv);
        break;
      
      default:
        console.error(chalk.red(`Unknown command: ${command}`));
        showHelp();
        process.exit(1);
    }
  } catch (error: any) {
    console.error(chalk.red('Error:'), error.message || error);
    process.exit(1);
  }
}

main();
