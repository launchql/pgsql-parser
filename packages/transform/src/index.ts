import { Node as PG13Node } from './13/types';
import { Node as PG14Node } from './14/types';
import { Node as PG15Node } from './15/types';
import { Node as PG16Node } from './16/types';
import { Node as PG17Node } from './17/types';

import * as PG13Types from './13/types';
import * as PG14Types from './14/types';
import * as PG15Types from './15/types';
import * as PG16Types from './16/types';
import * as PG17Types from './17/types';

export { ASTTransformer } from './transformer';
export { V13ToV14Transformer } from './transformers/v13-to-v14';
export { V14ToV15Transformer } from './transformers/v14-to-v15';
export { V15ToV16Transformer } from './transformers/v15-to-v16';
export { V16ToV17Transformer } from './transformers/v16-to-v17';
export { BaseTransformer, TransformerVisitor, TransformerContext } from './visitors/base';

export { PG13Node, PG14Node, PG15Node, PG16Node, PG17Node, PG13Types, PG14Types, PG15Types, PG16Types, PG17Types };
