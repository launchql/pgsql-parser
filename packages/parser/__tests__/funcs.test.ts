import { parseFunction } from '../src';
import { cleanTree } from '../src/utils';

const testFunction = `
CREATE FUNCTION ast.index_elem (
  v_name text DEFAULT NULL,
  v_ordering int DEFAULT NULL,
  v_nulls_ordering int DEFAULT NULL,
  v_expr jsonb DEFAULT NULL,
  v_opclass jsonb DEFAULT NULL,
  v_collation jsonb DEFAULT NULL
) RETURNS jsonb AS $EOFCODE$
DECLARE
  result jsonb = '{"IndexElem":{}}'::jsonb;
BEGIN
  result = ast.jsonb_set(result, '{IndexElem, name}', to_jsonb(v_name));
  result = ast.jsonb_set(result, '{IndexElem, ordering}', to_jsonb(v_ordering));
  result = ast.jsonb_set(result, '{IndexElem, nulls_ordering}', to_jsonb(v_nulls_ordering));
  result = ast.jsonb_set(result, '{IndexElem, expr}', v_expr);
  result = ast.jsonb_set(result, '{IndexElem, opclass}', v_opclass);
  result = ast.jsonb_set(result, '{IndexElem, collation}', v_collation);
  RETURN result;
END;
$EOFCODE$ LANGUAGE plpgsql IMMUTABLE;
     `;

it('works', async () => {
  const results = await parseFunction(testFunction);
  expect(cleanTree(results)).toMatchSnapshot();
});
