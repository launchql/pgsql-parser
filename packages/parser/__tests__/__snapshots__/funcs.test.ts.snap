// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
[
  {
    "PLpgSQL_function": {
      "action": {
        "PLpgSQL_stmt_block": {
          "body": [
            {
              "PLpgSQL_stmt_assign": {
                "expr": {
                  "PLpgSQL_expr": {
                    "query": "SELECT ast.jsonb_set(result, '{IndexElem, name}', to_jsonb(v_name))",
                  },
                },
                "lineno": 5,
                "varno": 7,
              },
            },
            {
              "PLpgSQL_stmt_assign": {
                "expr": {
                  "PLpgSQL_expr": {
                    "query": "SELECT ast.jsonb_set(result, '{IndexElem, ordering}', to_jsonb(v_ordering))",
                  },
                },
                "lineno": 6,
                "varno": 7,
              },
            },
            {
              "PLpgSQL_stmt_assign": {
                "expr": {
                  "PLpgSQL_expr": {
                    "query": "SELECT ast.jsonb_set(result, '{IndexElem, nulls_ordering}', to_jsonb(v_nulls_ordering))",
                  },
                },
                "lineno": 7,
                "varno": 7,
              },
            },
            {
              "PLpgSQL_stmt_assign": {
                "expr": {
                  "PLpgSQL_expr": {
                    "query": "SELECT ast.jsonb_set(result, '{IndexElem, expr}', v_expr)",
                  },
                },
                "lineno": 8,
                "varno": 7,
              },
            },
            {
              "PLpgSQL_stmt_assign": {
                "expr": {
                  "PLpgSQL_expr": {
                    "query": "SELECT ast.jsonb_set(result, '{IndexElem, opclass}', v_opclass)",
                  },
                },
                "lineno": 9,
                "varno": 7,
              },
            },
            {
              "PLpgSQL_stmt_assign": {
                "expr": {
                  "PLpgSQL_expr": {
                    "query": "SELECT ast.jsonb_set(result, '{IndexElem, collation}', v_collation)",
                  },
                },
                "lineno": 10,
                "varno": 7,
              },
            },
            {
              "PLpgSQL_stmt_return": {
                "expr": {
                  "PLpgSQL_expr": {
                    "query": "SELECT result",
                  },
                },
                "lineno": 11,
              },
            },
          ],
          "lineno": 4,
        },
      },
      "datums": [
        {
          "PLpgSQL_var": {
            "datatype": {
              "PLpgSQL_type": {
                "typname": "UNKNOWN",
              },
            },
            "refname": "v_name",
          },
        },
        {
          "PLpgSQL_var": {
            "datatype": {
              "PLpgSQL_type": {
                "typname": "UNKNOWN",
              },
            },
            "refname": "v_ordering",
          },
        },
        {
          "PLpgSQL_var": {
            "datatype": {
              "PLpgSQL_type": {
                "typname": "UNKNOWN",
              },
            },
            "refname": "v_nulls_ordering",
          },
        },
        {
          "PLpgSQL_var": {
            "datatype": {
              "PLpgSQL_type": {
                "typname": "UNKNOWN",
              },
            },
            "refname": "v_expr",
          },
        },
        {
          "PLpgSQL_var": {
            "datatype": {
              "PLpgSQL_type": {
                "typname": "UNKNOWN",
              },
            },
            "refname": "v_opclass",
          },
        },
        {
          "PLpgSQL_var": {
            "datatype": {
              "PLpgSQL_type": {
                "typname": "UNKNOWN",
              },
            },
            "refname": "v_collation",
          },
        },
        {
          "PLpgSQL_var": {
            "datatype": {
              "PLpgSQL_type": {
                "typname": "UNKNOWN",
              },
            },
            "refname": "found",
          },
        },
        {
          "PLpgSQL_var": {
            "datatype": {
              "PLpgSQL_type": {
                "typname": "jsonb ",
              },
            },
            "default_val": {
              "PLpgSQL_expr": {
                "query": "SELECT '{"IndexElem":{}}'::jsonb",
              },
            },
            "lineno": 3,
            "refname": "result",
          },
        },
      ],
    },
  },
]
`;
